### DISCUSSION TITLE
Nuxt 3 Layouts loading data

### DISCUSSION THREAD

**Discussion by memic84** [nuxt/nuxt#16167] [Questions] [answered]
Is it possible to load data in a nuxt layout from external sources. I am using SSR, and loading the data from a CMS.
I keep getting a node mismatch, because the data in the default layout is not populated until the client is read.
The problem seems to be the fact that the Layout is executed, and after that the Store and ...slug page etc.

This worked in Nuxt 2, but seems different now. I don't want to put the some of the layout components in the pages or components, because of repetition. Any ideas on how to fix this.

![image](https://user-images.githubusercontent.com/3043840/204581113-f4652eb5-f39c-434f-9cab-61877793799d.png)

**Comment by TheAlexLichter** [comment, marked as answer]
The problem is probably rather that you populate the breadcrumbs that exist in your layout when your page component was loaded. 

During SSR, components will be rendered top-down and won't be adapted then. Means, you cannot change e.g. components in the layout from a page component on server-side (through a store for example).

**Reply by memic84** [reply]
Hi @manniL, that's what i was afraid of. I believe that i have done it this way in Nuxt 2, is this part of SSR mechanism totally different?

I have a breadcrumb component in `default.vue` layout, which is populated from the `page [...slug].vue` from the root pages folder, and indeed from debugging i saw that the layout is of course first loaded, and after that the catchAll is done. But when using useFetch, should this be awaited, and rendered when all data is present? Of course this issue only happens when SSR kicks in, and not on subsequent load.

**What would be a good way to load data for a component in a layout, when using catchAll routes?** 
Moving breadcrumbs and other global components deeper, inside pages of components, would mean a lot of repetition.

More info:
- The catchAll route fetches data from /server/api/page -> which fetches data from a CMS
- Before the data is returned to server/api/page we're doing some data mapping from the CMS
- We're also using some custom pages from pages/, which aren't used in the catchAll route

**Reply by TheAlexLichter** [reply]
> Hi @manniL, that's what i was afraid of. I believe that i have done it this way in Nuxt 2, is this part of SSR mechanism totally different?

No, this should've led to hydration errors in Nuxt 2 already 🤔

> But when using useFetch, should this be awaited, and rendered when all data is present?

It will be awaited for the "component level" only. You can't await all at once because children might depend on data of their parent comopnents.

> What would be a good way to load data for a component in a layout, when using catchAll routes?

You can try using `definePageMeta` to resolve this. Define your breadcrumbs function there and call (+ await) it in the layout/`app.vue`.

**Reply by memic84** [reply]
>  No, this should've led to hydration errors in Nuxt 2 already 🤔

It's really weird, because Nuxt 2 doesn't show errors about hydration, nor do you see a Content Layout Shift, when loading the page. In the source the data is available in Nuxt 2. However, it does seem logical that this doesn't work, because we're loading the data, when it's not present yet.

The breadcrumbs data is returned from the same call as all other page properties. I have made a minimal reproduction:

https://stackblitz.com/edit/nuxt-starter-vgyw8s?file=server%2Fapi%2Fpage.ts,stores%2FusePageStore.ts,pages%2Fabout.vue,pages%2F[...slug].vue,layouts%2Fdefault.vue

I don't understand how i could implement the `definePageMeta` to send the data to the default layout data, or await it there.

It seems to me, that this could be a use-case that comes up a lot, when you're not data fetching inside your layout, but want to use the child component data.

**Reply by TheAlexLichter** [reply]
Example from one of my projects:

```html
<script setup lang="ts">
//layout or app.vue
const route = useRoute()
const breadcrumbs = computed(() => route.meta.breadcrumbs?.(route) ?? [])
</script>
```

```html
<script setup lang="ts">
//page
definePageMeta({
  breadcrumbs (route: RouteLocationNormalized) {
    return [
      {
        name: 'Category here',
        to: '/lalala/',
      }, {
        name: callSomeFn().name,
        to: route.path,
      }
    ]
  }
})
</script>
```

**Reply by memic84** [reply]
Thank you @manniL

**Reply by memic84** [reply]
@manniL This works for me, when the data is static and not being fetched from outside.

[...slug].vue
```vue
const { page } = storeToRefs(usePageStore())
await getPage(slug)

definePageMeta({
  breadcrumbs(route: RouteLocationNormalized) {
    return page?.value?.breadcrumbs
  },
})
```

I might have misunderstood this, but the `page` is not available in the definePageMeta. Is the definePageMeta executed before created() / init?

**Reply by memic84** [reply]
Another solution could be, to create a **wrapper/container component** for each page, which will hold the breadcrumb to avoid a lot of repetition, and use slots for the content.

components/ContainerPage.vue
```vue
<template>
  	<div class="container">
		<AppBreadcrumb :items='items'  />
     	<slot></slot>
	</div>
</template>
```

[...slug].vue
```vue
<template>
  	<ContainerPage>
		page specific content
	</ContainerPage>
</template>
```

**Comment by memic84** [comment]
Here is a minimal reproduction of the code: [https://stackblitz.com/edit/nuxt-starter-vgyw8s?file=server%2Fapi%2Fpage.ts,stores%2FusePageStore.ts,pages%2Fabout.vue,pages%2F[...slug].vue,layouts%2Fdefault.vue](https://stackblitz.com/edit/nuxt-starter-vgyw8s?file=server%2Fapi%2Fpage.ts,stores%2FusePageStore.ts,pages%2Fabout.vue,pages%2F%5B...slug%5D.vue,layouts%2Fdefault.vue)
